### PR TITLE
Enable op state pause and resume for compatible states

### LINKED ISSUE #30742: [1.3] RvcOpState - Enable pause and resume for compatible states
Following [this spec PR](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/7646), update the RvcOperationState cluster's handling of the pause and resume commands such that the states that are not pause-compatible or resume-compatible will return an InvalidInState error when receiving the pause or resume command respectively.

The SDK should allow all the pause and resume commands for vendor defined states to reach the delegate. It will be up to the delegate do decide if the command should be accepted or not.

Note that derived clusters share the same number space for defined states. This means that knowledge about the derived cluster is required when comparing derived cluster states.

### PR DESCRIPTION
**Pull request by hicklin**
Fixes #30742. 

Makes the private variables `mClusterId` and `mEndpointId` in the Operational State server Instance `const`. This may allow compilers to compile out code that is not reachable.

### CODE REVIEW

File: src/app/clusters/operational-state-server/operational-state-server.cpp [hunk 1]
```diff
@@ -395,11 +410,24 @@ void Instance::HandleResumeState(HandlerContext & ctx, const Commands::Resume::D
...
+    {
+    case RvcOperationalState::Id:
+        if (opState == to_underlying(RvcOperationalState::OperationalStateEnum::kSeekingCharger))
+        {
+            err.Set(to_underlying(ErrorStateEnum::kCommandInvalidInState));
+        }
+    }
+
+    // If the error is still NoError, we can call the delegate's handle function.
+    if (err.errorStateID == 0)
```
Comment (plauric) [inline src/app/clusters/operational-state-server/operational-state-server.cpp:430]: Should this call be skipped if in the Running opstate?

File: src/app/clusters/operational-state-server/operational-state-server.cpp [hunk 2]
```diff
@@ -337,11 +337,26 @@ void Instance::HandlePauseState(HandlerContext & ctx, const Commands::Pause::Dec
...
+    if (opState == to_underlying(OperationalStateEnum::kStopped) || opState == to_underlying(OperationalStateEnum::kError))
     {
         err.Set(to_underlying(ErrorStateEnum::kCommandInvalidInState));
     }
-    else if (opState == to_underlying(OperationalStateEnum::kRunning))
+
+    // Handle Pause-incompatible states for derived clusters.
+    switch (mClusterId)
+    {
+    case RvcOperationalState::Id:
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-state-server/operational-state-server.cpp:349]: This is almost certainly not going to be optimized out by compilers for instances of this thing that are not RvcOperationalState.  A compiler would have to prove that the class is never created anywhere with specific values of the cluster id, which is not something it knows in this translation unit.  The only thing that could make it work is if you're compiling with an LTO pass or something, but is that common for embedded devices?  Certainly our build system does not make it very feasible right now.

If we want this to really be "0-cost when not relevant", we should do one of two things:

1. We have a virtual method that subclasses that are cluster-id-specific need to implement to do this check.  Then the cost here is constant no matter how many opstate things we happen to have around, and this scales sanely to vendor-prefixed opstate derivations and so on.
2. We make the cluster id a template param.  But then you probably end up with some cost if you ever have more than one opstate instance around with different cluster ids, so this is likely not a good idea.

File: src/app/clusters/operational-state-server/operational-state-server.h [hunk 3]
```diff
@@ -174,6 +174,9 @@ class Instance : public CommandHandlerInterface, public AttributeAccessInterface
 
     /**
      * Handle Command: Pause.
+     * If the current state is not pause-compatible, this method responds with an ErrorStateId of CommandInvalidInState.
+     * If the current state is pause, this method responds with an ErrorStateId of NoError but takes no action.
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-state-server/operational-state-server.h:178]: There is no state named "pause".

Comment (hicklin) [reply src/app/clusters/operational-state-server/operational-state-server.h]: Fixed.

File: src/app/clusters/operational-state-server/operational-state-server.cpp (hunk 2 above)
Comment (hicklin) [reply src/app/clusters/operational-state-server/operational-state-server.cpp]: I have changed the API such that all operational state clusters implement their own class. If required, they can implement additional logic by overriding virtual protected functions of the base operational state class.

**Review by bzbarsky-apple — APPROVED** [review]
Much better, thank you!

File: src/app/clusters/operational-state-server/operational-state-server.h [hunk 4]
```diff
@@ -275,6 +312,66 @@ class Delegate
...
+     * Note: if a state outside the derived cluster number-space is given, this method returns false.
+     * @param aState The state to check.
+     * @return true if aState is pause-compatible, false otherwise.
+     */
+    bool IsDerivedClusterStateResumeCompatible(uint8_t aState) override;
+};
+
+} // namespace RvcOperationalState
+
+namespace OvenCavityOperationalState {
```
Comment (plauric) [inline src/app/clusters/operational-state-server/operational-state-server.h:354]: What's the purpose of this namespace?

File: src/app/clusters/operational-state-server/operational-state-server.h [hunk 5]
```diff
@@ -275,6 +312,66 @@ class Delegate
...
+
+class Instance : public OperationalState::Instance
+{
+public:
+    /**
+     * Creates an RVC operational state cluster instance.
```
Comment (plauric) [inline src/app/clusters/operational-state-server/operational-state-server.h:360]: Should this match the namespace's name?

File: src/app/clusters/operational-state-server/operational-state-server.h (hunk 4 above)
Comment (hicklin) [reply src/app/clusters/operational-state-server/operational-state-server.h:354]: The API has changed. The need for the oven namespaace is to create an instance of the oven Operational State cluster with the correct cluster ID.

**Review by plauric — APPROVED** [review]

**Pull request merged**